### ISSUE TITLE
Figures not properly closed when using plt.close() in python script

### ISSUE THREAD

**Issue by tfetherolf** [matplotlib/matplotlib#18644]
### Bug report

This bug was originally appended to issue #17109 and suggested to be submitted as a new issue by @jklymak. 

**Bug summary**

When creating several figures in a script that is run from the command line, the figures do not seem to properly close when calling `plt.close()`. There are no warnings shown, but creating a large number of figures in a loop results in an error stating that too many files are open.

`OSError: [Errno 24] Too many open files: 'figurename.png'`

All open files/figures have been closed within each loop and "garbage" is regularly collected within the loop. 

```
import gc

for i in range(1000):
    if i % 50:
        gc.collect()
```

Increasing or removing the limit on the number of open files allowed is not an adequate solution for me since this would cause extreme CPU slowdowns from the large number of figures that I am creating.

**Code for Reproduction**

Based on the a fix that I found, I suspect that this error could be reproduced by running the following in a .py file. 

```
import matplotlib.pyplot as plt

for i in range(1000):
    plt.figure()
    plt.plot(1, 1, 'ko')
    plt.savefig('figurename'+str(i)+'.png')
    plt.close()
```

In my code the error occurs shortly before iteration 256 because a few other files are also open in each loop iteration. Unfortunately I am unable to actually reproduce the error when using this code snippet separately, but I'm not sure why besides the degree of information that is stored in the actual figures in my more complicated code... The error appears when trying to save the figure and is resolved when avoiding `plt.close()`, as demonstrated below. 

**Current Workaround**

I found that opening a figure before the loop and using plt.clf() instead of opening/closing many figures is an adequate workaround in my case.

```
import matplotlib.pyplot as plt

plt.figure()
for i in range(1000):
    plt.plot(1, 1, 'ko')
    plt.savefig('figurename'+str(i)+'.png')
    plt.clf()
```

**Expected outcome**

The code using `plt.figure()` and `plt.close()` within large loops used to work in a python3 conda environment that had older versions of most packages (including matplotlib). Unfortunately, the working version information is unavailable because this computer recently had several parts replaced resulting in complete storage loss and a fresh OS install. 

**Matplotlib version**

  * Operating system: macOS Catalina 10.15.7
  * Matplotlib version: 3.31
  * Matplotlib backend: MacOSX
  * Python version: 3.8.5
  * Other libraries: 
     * Qt version: 5.9.7
     * PyQt5 version: 5.9.2

I installed the command line version of anaconda3 and all packages were updated with `conda update --all`. Using `which python` confirms that my command line version is connected to anaconda. 

The hardware on this computer was recently replaced, so everything has been cleanly installed in the last few days. I also encountered these issues on a loaner macOS with everything similarly freshly installed, so it is not related to the computer parts being replaced in this model.

**Comment by timhoffm** [comment]
Without a minimal working example it's hard to tell what is going wrong. This may also be OS dependent. As to why this happens now and not in earlier versions of Matplotlib, we've lately been switching to pillow for writing pngs to file - not sure if that's related.

**Comment by jklymak** [comment]
Yeah, when I asked you to post this, I didn't note that the examples you gave don't trigger the error.  We'd need something reproducible to help.  Perhaps your script is using threading?  Matplotlib is not thread safe.

**Comment by tfetherolf** [comment]
My script is not using threading. I tried fiddling with the minimum working example a bit more to force a crash situation, but I can't seem to figure out the magic combination between my larger code and the working example. I will close this issue for now, but will update here if I find out how to reproduce the error using a simple example in the future.

**Comment by jklymak** [comment]
Sorry its still causing problems.  Don't hesitate to re-open if you track it down.

**Comment by HotDog702** [comment]
I confirm the issue. I have the similar scheme: create and close figures in a cycle, Spyder, Qt5 backend, Ubuntu
After several runs, the figures remains in zombie state: I see a lot of window frames with transparent content, that can be closed only manually by pressing [x]. 
_plt.get_figlabels()_ or _plt.get_fignums()_ do not see them
The code from @tfetherolf reproduces the situation, but **IF** it already happens (_in the other code_). I encountered this bug in my big code, opened **a new console** - and voila - the above short cycle reproduces a bug! (but not all windows becomes zombie: I've got only half of them)
After restarting the computer, the bug disappears for a while. Tried to reproduce it again - but succeeded only once (( 

What I understand, the issue happens when you have more than 1 open window, some hard plotting task (like array of 100000 points) and actively switch between windows while pyplot do its job

**Comment by tacaswell** [comment]
Do you need to see the plots?  If so, then you need to run the event loop for at least some period of time (`plt.pause(...)` or manually flushing the events), if not then use a non-interactive backend.

See 
https://matplotlib.org/stable/users/explain/figure/interactive_guide.html#explicitly-spinning-the-event-loop

**Comment by HotDog702** [comment]
> Do you need to see the plots?

It depends... I definitely need not see figures during plotting, but... If I process one file - I do not call **plt.close(..)** to analyze plots after the script is finished, and if I process a set of files - I save plots to the file and close them.
Could I define backend when invoking **plt.figure()** ? I want to use it for certain figures, not globally

**Comment by jklymak** [comment]
This will remain a mystery without a reproducible example.  Also suggest removing PyCharm from the debugging if possible. If it still fails without pycharm feel feee to open a new issue.  If it only happens with pycharm perhaps report to them?

**Comment by jklymak** [comment]
This will remain a mystery without a reproducible example.  Also suggest removing PyCharm from the debugging if possible. If it still fails without pycharm feel feee to open a new issue.  If it only happens with pycharm perhaps report to them?

**Comment by HotDog702** [comment]
> If it only happens with pycharm perhaps report to them?

I use Spyder, not PyCharm

**Comment by HotDog702** [comment]
By the way, it may be some cumulative effect, as it disappears after restarting the computer (not checked restarting the IDE, sorry)
